### ISSUE TITLE
Web: improve consumption of our web bits as an embedder

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#125643]
We provide an API surface for an embedder of our web UI:

https://github.com/microsoft/vscode/blob/7a83e5aa6278d27120c26dabe2b59642dcad9cf2/src/vs/workbench/workbench.web.api.ts#L238-L238

We also require some specific setup to load our web UI pieces as indicated by our `yarn web` implementation:

https://github.com/microsoft/vscode/blob/c87c95a1a04b01107328fe1987b1865b2eb367a8/src/vs/code/browser/workbench/workbench.html#L42-L42

**We should:**
* investigate to produce a `workbench.web.api.d.ts` (similar to `monaco.d.ts`) for the `workbench.web.api` for an embedder to develop against
* consider to extract the loader configuration into something reusable (`amdconfig.js`?) that an embedder can just point to

**Comment by lramos15** [comment]
Loader piece is done will defer the .d.ts to a later date

**Comment by connor4312** [comment]
I suggest _not_ automatically generating it, and instead have the web api .d.ts checked into VS Code and with manually duplicated types.

The reason is that we have dependency on code and typings for embedders, but it is very easy for team members to not realize that code they're working is exposed in this API and depended upon by the embedder. This may result in breakages, especially if it's something as easy and innocent as an automated refactor for a property rename. Causing a compilation error unless the web .d.ts is manually and explicitly updated would reduce the likelihood of accidental breakages.

**Comment by bpasero** [comment]
I think there are 2 camps: `vscode.d.ts` is handcrafted but `monaco.d.ts` for example is not and both have API users.

**Comment by jrieken** [comment]
> I suggest not automatically generating the .d.ts, and instead have the web api .d.ts checked into VS Code and with manually duplicated types.

+💯. Creating those types manually is IMO the only way to guarantee awareness of changes in API and to "design" the API explicitly. That's also the different between `vscode.d.ts` and `monaco.d.ts` - the latter is a slightly cherry-picked but mostly AS-IS representation of the "internal" API.

**Comment by bpasero** [comment]
I am fine handpicking the API for embedder but I wonder how we could validate the correctness of the API then: we need to ensure we implement the API also in the workbench to get type failures early. In other words, we cannot just hand craft a d.ts file and not validate it with our actual code. I think `vscode.d.ts` does that.

**Comment by jrieken** [comment]
Yeah, we explicitly implement the API in `src/vs/workbench/api/common/extHost.api.impl.ts`, we define a type which is of the `vscode`-module here: https://github.com/microsoft/vscode/blob/527d63162ad6b43a5e106e93ed05c3a299bda947/src/vs/workbench/api/common/extHost.api.impl.ts#L1138

**Comment by bpasero** [comment]
@joaomoreno pointed out this tool which maybe could help: https://api-extractor.com/

**Comment by samdenty** [comment]
@bpasero i'm interested in this, any progress?

**Comment by samdenty** [comment]
I tried using api-extractor by using the new typescript 5.6.0 noCheck option to compile a separate folder of `.d.ts` for the src folder in hopes stripping out the code causing the error and just using the types would help. But after i did that i ran into a new error `Child declaration not found for the specified node`, it was happening on pretty much any import from `vs/`.

Then I tried using https://github.com/timocov/dts-bundle-generator and it worked a charm, I created:

declarations/package.json:

```json
{
  "name": "declaration",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "overrides": {},
  "dependencies": {
    "dts-bundle-generator": "^9.5.1",
    "typescript": "5.6.0-dev.20240723"
  },
  "resolutions": {
    "**/typescript": "5.6.0-dev.20240723"
  },
  "scripts": {
    "build": "dts-bundle-generator -o workbench.d.ts --project tsconfig.json ../src/vs/workbench/browser/web.api.ts"
  }
}
```

declarations/tsconfig.json:

```json
{
	"compilerOptions": {
		"moduleResolution": "node",
		"baseUrl": "../src",
		"paths": {
			"vs/*": ["../src/vs/*"]
		},
		"target": "es2022",
		"skipLibCheck": true,
		"noCheck": true,
		"lib": ["ES2022", "DOM", "DOM.Iterable", "WebWorker.ImportScripts"]
	},
	"include": [
		"../src/vs/**/*.ts",
		"vscode-dts/vscode.proposed.*.d.ts",
		"vscode-dts/vscode.d.ts"
	]
}
```

and then run `yarn build` and i get a nicely compiled `workbench.d.ts`

**Comment by bpasero** [comment]
Cool, can you attach the resulting file here? Curious how it compares to the one we manually craft.

**Comment by bpasero** [comment]
Btw when I try I get an error:

```
~/Development/Microsoft/vscode/declarations> yarn build
yarn run v1.22.22
$ dts-bundle-generator -o workbench.d.ts --project tsconfig.json ../src/vs/workbench/browser/web.api.ts
Compiling input files...
error TS5023: Unknown compiler option 'noCheck'.

Error: Error while processing tsconfig compiler options
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
~/Development/Microsoft/vscode/declarations>   
```
